WordPress/gutenberg
# [Mobile] List V2 - Fixes split issues

Fixes https://github.com/WordPress/gutenberg/issues/43873

## What?
This PR fixes some split issues like trying to split a List (V2) in the middle or merging a Paragraph block and a List Item block.

## Why?
To have the same functionality as the web editor and the previous version of the List.

## How?
It adds a [native variant](https://github.com/WordPress/gutenberg/pull/43949/commits/93212f24bdd410c19e2ed6da2a53ec97b3126d51) of the `useEnter` hook modified for native, on the web editor it uses listeners for the `onKeyPress` where it also can stop other events to trigger in some specific scenarios so we couldn't have the same functionality on native.

So on native we are now passing an `onEnter` prop to the `RichText` component where we can handle the custom `onSplit` functionality for the List block.

It also cherry-picks https://github.com/WordPress/gutenberg/pull/42780 as it is needed to be able to merge for example a Paragraph block with a List Item block.

## Testing Instructions

- Insert a List block with the native editor.
- Add at least three items to the list.
- Place the cursor at the end of one of the items in the middle of the list.
- Press the `Return` key twice.
- **Expect** to see a Paragraph block added.
- Type something in this new block.
- Move the cursor to the start of the text.
- Tap on the `Backspace` key.
- **Expect** the Paragraph block to be transformed into a List item block.

## Screenshots or screencast 

Before iOS | After iOS
-|-
<kbd><video src="https://user-images.githubusercontent.com/4885740/188877560-fb7cfd9e-efc8-447f-906f-913b74a7b608.mov" width="200" /></kbd> | <kbd><video src="https://user-images.githubusercontent.com/4885740/188877798-ff6e7bda-b8ea-4231-9d00-13082f230d7f.mov" width="200" /></kbd>

Before Android | After Android
-|-
<kbd><video src="https://user-images.githubusercontent.com/4885740/188877908-c3e4a82b-fd92-4503-a8fa-71726e902e1b.mov" width="200" /></kbd> | <kbd><video src="https://user-images.githubusercontent.com/4885740/188878048-d9ca556b-6504-48c0-9d05-bdf6188c46af.mov" width="200" /></kbd>
